go/consensus/tendermint/abci/state.go

```diff
@@ -421,78 +331,140 @@ func (s *applicationState) metricsWorker() {
 
 	for {
 		select {
-		case <-s.metricsCloseCh:
+		case <-s.ctx.Done():
 			return
 		case <-t.C:
 			_ = s.updateMetrics()
 		}
 	}
 }
 
+func (s *applicationState) pruneWorker() {
+	defer close(s.prunerClosedCh)
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case r := <-s.prunerNotifyCh.Out():
+			round := r.(uint64)
+
+			if err := s.statePruner.Prune(s.ctx, round); err != nil {
+				s.logger.Warn("failed to prune state",
+					"err", err,
+					"block_height", round,
+				)
+			}
+		}
+	}
+}
+
 func newApplicationState(ctx context.Context, cfg *ApplicationConfig) (*applicationState, error) {
-	db, err := db.New(filepath.Join(cfg.DataDir, "abci-mux-state"), false)
+	baseDir := filepath.Join(cfg.DataDir, appStateDir)
+	if err := common.Mkdir(baseDir); err != nil {
+		return nil, fmt.Errorf("failed to create application state directory: %w", err)
+	}
+
+	switch cfg.StorageBackend {
+	case storageDB.BackendNameBadgerDB:
+	default:
+		return nil, fmt.Errorf("unsupported storage backend: %s", cfg.StorageBackend)
+	}
+
+	db, err := storageDB.New(&storage.Config{
+		Backend:          cfg.StorageBackend,
+		DB:               filepath.Join(baseDir, storageDB.DefaultFileName(cfg.StorageBackend)),
+		MaxCacheSize:     64 * 1024 * 1024, // TODO: Make this configurable.
+		DiscardWriteLogs: true,
+		NoFsync:          true, // This is safe as Tendermint will replay on crash.
```

Comment:
So in case the kernel/machine crashes while performing a commit, any Badger transactions are rolled back. The last thing that a `Finalize` does is to run a transaction that updates the last finalized round. In case that transaction is either not executed or is rolled back due to WAL corruption, when the node starts up again, `GetLatestRound` will return the previous round.

This will cause Tendermint (assuming that Tendermint's WAL implementation is correct) to see that the application is behind and will replay the block that was not committed or partially committed. So all updates will happen again and finalize will be called again.

Assuming that all state updates are deterministic, this is safe. In the worst case (only if the updates are non-deterministic) there would be some leftovers in the database which would not be garbage collected (this could happen even if fsync is enabled).

Reply:
ok so NoFsync won't cause inconsistency